### PR TITLE
Nf core dev / variant calling bugs / ideas

### PR DESCRIPTION
**Pull request by nickhsmith**
## PR checklist

The following changes were made after trying to run the variant_calling pipeline (specifically ` --tools haplotypecaller`). I have readded the `gatk4/genotypegvcfs` module from nf-core (which was updated to include interval index files)

### An overview of the changes file by file:
`concatenateVCFs.sh` : The default for bcftools sorting is on the /tmp directory (or /scratch) however it makes sense to me to maintain the temp.sorting files in the location of your work directory (`.`)

`conf/base.config`: GENOTYPEGVCFS is listed as a medium_priority tool, but for WGS it was constantly failing the first 1 or 2 attempts. Giving it the default `high_priority` mem/cpus

`conf/modules.config`: added the GENOTYPEGVCFS module to publish the final vcf file. Edited the HAPLOTYPECALLER module to not publish the gvcfs by default (could be set to true, but they are huge for a WGS sample) and to name them "{meta.id}.g" to differentiate them from the final vcf file.

`lib/WorkflowMain.groovy`: the nextflow pipeline requires an input file. But this is in direct contrast to the automatic restart based on the `results/../.csv` used in sarek.nf to determine the step and input files. Not elegant, but works to restart if appropriate step is named

`subworkflows/local/germline_variant_calling.nf`: include GENOTYPEGVCFS and account for both with intervals genotyping (work on the gvcfs after concat) or no intervals (work on the gvfs without concat) and create a new channel for the GENOTYPEGVCFS input. Should eventually be linked with logic for joint variant calling. The current behaviour is ALWAYS do single sample variant calling, and if joint variant calling is called, do that also. (not ideal)

### CODE REVIEW

**Review by maxulysse — APPROVED** [review]

**Comment by maxulysse** [comment]
> `concatenateVCFs.sh` : The default for bcftools sorting is on the /tmp directory (or /scratch) however it makes sense to me to maintain the temp.sorting files in the location of your work directory (`.`)

I'd rather keep the default for `bcftools` or change the default. 

> `conf/base.config`: GENOTYPEGVCFS is listed as a medium_priority tool, but for WGS it was constantly failing the first 1 or 2 attempts. Giving it the default `high_priority` mem/cpus

Why not changing the modules to a `high_priority` tool?

> `conf/modules.config`: added the GENOTYPEGVCFS module to publish the final vcf file. Edited the HAPLOTYPECALLER module to not publish the gvcfs by default (could be set to true, but they are huge for a WGS sample) and to name them "{meta.id}.g" to differentiate them from the final vcf file.

Sounds like a good idea

> `lib/WorkflowMain.groovy`: the nextflow pipeline requires an input file. But this is in direct contrast to the automatic restart based on the `results/../.csv` used in sarek.nf to determine the step and input files. Not elegant, but works to restart if appropriate step is named

Not seeing any other solution there for now

> `subworkflows/local/germline_variant_calling.nf`: include GENOTYPEGVCFS and account for both with intervals genotyping (work on the gvcfs after concat) or no intervals (work on the gvfs without concat) and create a new channel for the GENOTYPEGVCFS input. Should eventually be linked with logic for joint variant calling. The current behaviour is ALWAYS do single sample variant calling, and if joint variant calling is called, do that also. (not ideal)

I see what you mean there

**Comment by FriederikeHanssen** [comment]
> `concatenateVCFs.sh` : The default for bcftools sorting is on the /tmp directory (or /scratch) however it makes sense to me to maintain the temp.sorting files in the location of your work directory (`.`)

I don't understand the advantages of this. can you elaborate?

> `conf/base.config`: GENOTYPEGVCFS is listed as a medium_priority tool, but for WGS it was constantly failing the first 1 or 2 attempts. Giving it the default `high_priority` mem/cpus

I agree that this should be changed upstream


> `lib/WorkflowMain.groovy`: the nextflow pipeline requires an input file. But this is in direct contrast to the automatic restart based on the `results/../.csv` used in sarek.nf to determine the step and input files. Not elegant, but works to restart if appropriate step is named

not sure about a solution, but also haven't given it any thought. Honestly, not very happy about changing the temaplte code. Can we maybe take this disucssion to #tools before making any changes here? Looks like a release there will happen soon anyways, so it would maybe be good to get it in there now

> `subworkflows/local/germline_variant_calling.nf`: include GENOTYPEGVCFS and account for both with intervals genotyping (work on the gvcfs after concat) or no intervals (work on the gvfs without concat) and create a new channel for the GENOTYPEGVCFS input. Should eventually be linked with logic for joint variant calling. The current behaviour is ALWAYS do single sample variant calling, and if joint variant calling is called, do that also. (not ideal)

Joint variant calling subworkflow needs a bit of work still anyways. So fine with merging it here and then work on the changes in a separate PR.

**Comment by nickhsmith** [comment]
the bcftools sort in the `concatenateVCFs.sh` file is used because the concatenation of all of the intervals (by chromosome) builds a vcf that is unsorted. bcftools by default uses the `/tmp/` dir to create temporary `/tmp/bcftools-sort.XXXXXX` files that are then merged to create the sorted vcf file.

If a user is on a shared file system that doesn't have write access to `/tmp/` or if this is not very large (for WGS these temp gvcf files can get quite big) then you get errors. By changing the default temp space to the local directory `.` which is actually in the `work/hash1/hash2/` dir we make sure that the user has write access to this location, and that presumably the `work_dir` is large enough to handle all of the the intermediate files this negates the problems mentioned above (although at the cost of potential I/O depending on how `/tmp/` and `/PATH/TO/WORK/` storage is built)

**Comment by nickhsmith** [comment]
I agree that we can change the upstream priority for `nf-core/modules/genotypegvfs` but that makes assumptions for other users that may not be ideal. I will make a PR for that though

**Comment by FriederikeHanssen** [comment]
ah now I see. Yes we had the same issue with GATK tools in the past and had to add `--TMP-DIR` or so to also allow to write to defined scratch space with the nextflow parameter `scratch = true` or `scratch = /some/place/where/intermediate/file/are/stored` . Did you try the process with setting the scratch directory as well to make sure that would work to? It would be preferabe to not enforce the work dir to be used for intermediate files, since it can slow down execution. I don't think your solution does, but just be sure maybe a quick test would be nice. Scratch directory for any process can be set in the custom.config too

**Comment by FriederikeHanssen** [comment]
> I agree that we can change the upstream priority for `nf-core/modules/genotypegvfs` but that makes assumptions for other users that may not be ideal. I will make a PR for that though

oretty sure we are the only ones using it at the moment :D If the raredisease pipeline needs it they would run into the same problem with WGS data.

**Comment by maxulysse** [comment]
> > I agree that we can change the upstream priority for `nf-core/modules/genotypegvfs` but that makes assumptions for other users that may not be ideal. I will make a PR for that though
> 
> oretty sure we are the only ones using it at the moment :D If the raredisease pipeline needs it they would run into the same problem with WGS data.

I think default should be more, so that it's easier to restrict when needed.
If default is less, then we really need to be sure to add more resources

**Comment by nickhsmith** [comment]
> TMP

What about a params.scratchdir so that we can set any scratch directive with process.scratch = "${params.scratchdir}"

**Comment by FriederikeHanssen** [comment]
this already exists to provide in a custom config/institutional config etc.. not sure, if process configurations should be command line parameters

**Comment by FriederikeHanssen** [comment]
yep it can also be set globally for all pipelines like here: https://github.com/nf-core/configs/blob/91b5875b44b01edfd2992e19adec0d1b552feef8/conf/cfc.config#L13 or just for your pipeline run for all processes:

```
process{
   scratch = false (writes to work/ then)
   
   or
   
   scratch = /some/directory/where/there/is/space
 }
```

**Comment by maxulysse** [comment]
Any comments on this module/update of ASCAT @malinlarsson ?

**Comment by FriederikeHanssen** [comment]
I am confused. What do you mean :D Lasse Folkersen is implementing the module right now, so I can add it to the subworkflow soon.

**Comment by maxulysse** [comment]
> I am confused. What do you mean :D Lasse Folkersen is implementing the module right now, so I can add it to the subworkflow soon.

Sorry, replied in the wrong PR...

File: subworkflows/local/germline_variant_calling.nf [hunk 1]
```diff
@@ -62,6 +63,7 @@ workflow GERMLINE_VARIANT_CALLING {
     deepvariant_vcf       = Channel.empty()
     freebayes_vcf         = Channel.empty()
     haplotypecaller_gvcf  = Channel.empty()
+    genotypegvcfs_vcf      = Channel.empty()
```
Comment (maxulysse) [inline subworkflows/local/germline_variant_calling.nf:66]:
```suggestion
    genotypegvcfs_vcf     = Channel.empty()
```

**Review by maxulysse — APPROVED** [review]

File: .gitignore [hunk 2]
```diff
@@ -7,3 +7,4 @@ results/
 testing/
 testing*
 *.pyc
+*.swp
```
Comment (FriederikeHanssen) [inline .gitignore:10]:
```suggestion
```

File: lib/WorkflowMain.groovy [hunk 3]
```diff
@@ -71,12 +71,6 @@ class WorkflowMain {
 
         // Check AWS batch settings
         NfcoreTemplate.awsBatch(workflow, params)
```
Comment (maxulysse) [inline lib/WorkflowMain.groovy:73]:
```suggestion
        NfcoreTemplate.awsBatch(workflow, params)

        // Check input has been provided
        if (!params.input) {
            log.error "Please provide an input samplesheet to the pipeline e.g. '--input samplesheet.csv'"
            System.exit(1)
        }
```

Comment (maxulysse) [reply lib/WorkflowMain.groovy:73]: Let's revert that for now as it need fixing somewhere else

File: conf/base.config [hunk 4]
```diff
@@ -60,6 +60,12 @@ process {
         memory = { check_max( 60.GB * task.attempt, 'memory' ) }
         time   = { check_max( 48.h  * task.attempt, 'time'    ) }
     }
+    withName:GATK4_GENOTYPEGVCFS {
+        // override process_medium tag parameters with the high_process tag specs
+        cpus   = { check_max( 12    * task.attempt, 'cpus'    ) }
+        memory = { check_max( 72.GB * task.attempt, 'memory'  ) }
+        time   = { check_max( 16.h  * task.attempt, 'time'    ) }
+    }
```
Comment (maxulysse) [inline conf/base.config:68]:
```suggestion
```

**Pull request merged**